WordPress/gutenberg
# [RNMobile] Fix bundling errors with metro

## Description


The errors while bundling react-native-editor on CI were fixed with this PR: https://github.com/WordPress/gutenberg/pull/23452

What it did was:
1. Install `watchman` before bundling
1. Remove a patch for `metro` that we thought wasn't needed for version 0.58.0

But the `metro` version used while bundling was actually 0.56.4 as it was not hoisted.
We also had problems that `watchman` was still needed while bundling locally.

To solve these problems this PR:

1. Uses `patch-package` to patch `metro` pre-bundling and reverts it post-bundling (we still need `metro` to watch files when it's running in server mode)
1. Removes the need to install `watchman` on CI or locally while bundling (thanks to this patch)
1. Extra: Now that we have `patch-package` dependency in `gutenberg`, also moves a `react-native` patch that was in `gutenberg-mobile` to `gutenberg`, so that the demo app in `gutenberg` also works as expected.

Related gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2474/

## How has this been tested?




On CI:
1. `clean` and `bundle` 25 times using this script:

```for i in $(seq 1 25); do npm run native clean && npm run native clean:runtime && npm run native test:e2e:bundle:ios zsh -i -c exit; done```

2. revert this change and run again x2

Locally:
1. Uninstall watchman (`brew uninstall watchman`)
1. Run the same script above
1. Run the metro server with `npm run native start:reset`
1. Check if iOS/Android demo apps are loading correctly from the server
1. Re-install watchman (`brew install watchman`)
1. Run the metro server again
1. Check if iOS/Android demo apps are loading correctly from the server

To see if `react-native` patch is applied:
1. Run `npm install` or `npm ci`
1. Go to `node_modules/react-native/React/Views/RCTShadowView.m` and check if there is this code:
`
while (shadowView && shadowView != ancestor) {
`
1. Go to `node_modules/react-native/Libraries/ActionSheetIOS/RCTActionSheetManager.m` and check if there is this code:
`if ([disabledButtonIndices containsObject:@(localIndex)]) {`
